url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/C64Screen.java

```diff
@@ -0,0 +1,2025 @@
+          setIRQ(VIC_IRQ);
+          lastIRQ = cpu.cycles;
+          if (IRQDEBUG)
+            monitor.info("Generating IRQ at " + vbeam + " req:" + raster
+                + " IRQs:" + cpu.interruptInExec
+                + " flags: " + irqFlags + " delta: " +
+                (cpu.cycles - lastLine));
+        }
+      } else {
+        irqTriggered = false;
+      }
+      notVisible = false;
+      if (vPos < 0 || vPos >= 284) {
+        cpu.baLowUntil = 0;
+        notVisible = true;
+        if (STATE_DEBUG)
+          monitor.info("FINISH next at " + vbeam);
+        // Jump directly to VS_FINISH and wait for end of line...
+        break;
+      }
+
+      // Check if display should be enabled...
+      if (vbeam == 0x30) {
+        displayEnabled = (control1 & 0x10) != 0;
+        if (displayEnabled) {
+          borderState &= ~0x04;
+        } else {
+          borderState |= 0x04;
+        }
+      }
+
+      badLine =
+        (displayEnabled && vbeam >= 0x30 && vbeam <= 0xf7) &&
+        (vbeam & 0x7) == vScroll;
+
+      // Clear the collission masks each line... - not needed???
+      for (int i = 0, n = SC_WIDTH; i < n; i++) {
+        collissionMask[i] = 0;
+      }
+      break;
+    case 1: // Sprite data - sprite 3
+      if (sprites[3].dma) {
+        sprites[3].readSpriteData(); // reads all 3 bytes here (one should be prev).
+      }
+      if (sprites[5].dma) {
+        cpu.baLowUntil = lastLine + VICConstants.BA_SP5;
+      }
+      break;
+    case 2:
+      // here some of the bytes for sprite 4 should be read...
+      break;
+    case 3:
+      if (sprites[4].dma) {
+        sprites[4].readSpriteData();
+      }
+      if (sprites[6].dma) {
+        cpu.baLowUntil = lastLine + VICConstants.BA_SP6;
+      }
+      break;
+    case 4:
+      // here some of the bytes for sprite 5 should be read...
+      break;
+    case 5:
+      if (sprites[5].dma) {
+        sprites[5].readSpriteData();
+      }
+      if (sprites[7].dma) {
+        cpu.baLowUntil = lastLine + VICConstants.BA_SP7;
+      }
+      break;
+    case 6:
+      // here some of the bytes for sprite 6 should be read..
+      break;
+    case 7:
+      if (sprites[6].dma) {
+        sprites[6].readSpriteData();
+      }
+      break;
+    case 8:
+      // here some of the bytes for sprite 7 should be read...
+      break;
+    case 9:
+      if (sprites[7].dma) {
+        sprites[7].readSpriteData();
+      }
+
+      // Border management! (at another cycle maybe?)
+      if (blankRow) {
+        if (vbeam == 247) {
+          borderState |= 1;
+        }
+      } else {
+        if (vbeam == 251) {
+          borderState |= 1;
+        }
+        if (vbeam == 51) {
+          borderState &= 0xfe;
+
+          // Reset sprite data to avoid garbage since they are not painted...
+          for (int i = 0, n = 7; i < n; i++) {
+            if (!sprites[i].painting) {
+              sprites[i].lineFinished = true;
+            }
+          }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1193:1202, 1205:1213.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.